internal/memory/pool.go

```diff
@@ -0,0 +1,324 @@
+// space.
+//
+// There are a total of 7 different region sizes that can be allocated, with the smallest
+// being 1MB and the largest 4GB (the default maximum size of a Virtual PMem device).
+//
+// For efficiency and to reduce fragmentation an entire region is allocated when requested.
+// When there's no available region of requested size, we try to allocate more memory for
+// this particular size by splitting the next available larger region into smaller ones, e.g.
+// if there's no region available for size class 0, we try splitting a region from class 1,
+// then class 2 etc, until we are able to do so or hit the upper limit.
+type PoolAllocator struct {
+	pools [memoryClassNumber]*memoryPool
+}
+
+var _ MappedRegion = &region{}
+var _ Allocator = &PoolAllocator{}
+
+func (r *region) Offset() uint64 {
+	return r.offset
+}
+
+func (r *region) Size() uint64 {
+	sz, _ := GetMemoryClassSize(r.class)
+	return sz
+}
+
+func newEmptyMemoryPool() *memoryPool {
+	return &memoryPool{
+		free: make(map[uint64]*region),
+		busy: make(map[uint64]*region),
+	}
+}
+
+func NewPoolMemoryAllocator() PoolAllocator {
+	pa := PoolAllocator{}
+	p := newEmptyMemoryPool()
+	// by default we allocate a single region with maximum possible size (class type)
+	p.free[0] = &region{
+		class:  memoryClassNumber - 1,
+		offset: 0,
+	}
+	pa.pools[memoryClassNumber-1] = p
+	return pa
+}
+
+// FindFreeOffset finds free memory region with minimum offset that can fit a given memory class.
+func (pa *PoolAllocator) FindFreeOffset(memCls classType) (uint64, error) {
+	if memCls >= memoryClassNumber {
+		return 0, ErrInvalidMemoryClass
+	}
+
+	for i := memCls; i < memoryClassNumber; i++ {
+		pi := pa.pools[i]
+		if pi == nil || len(pi.free) == 0 {
+			continue
+		}
+
+		// start off with max offset
+		target := 4 * GigaByte
+		for offset := range pi.free {
+			if offset < target {
+				target = offset
+			}
+		}
+		return target, nil
+	}
+	return 0, ErrNotEnoughSpace
+}
+
+// Allocate checks memory pool of class `memCls` and returns a free region with minimal offset, if none available,
+// tries to expanding current `memCls` pool.
+//
+// Internally it's done via moving a region from free pool into a busy pool
+func (pa *PoolAllocator) Allocate(memCls classType) (MappedRegion, error) {
+	if memCls >= memoryClassNumber {
+		return nil, ErrInvalidMemoryClass
+	}
+
+	// helper function to allocate a region
+	alloc := func() *region {
+		offset, err := pa.FindFreeOffset(memCls)
+		if err != nil {
+			return nil
+		}
+
+		if n, err := pa.swap(memCls, offset); err != nil {
+			return nil
+		} else {
+			return n
+		}
+	}
+
+	// allocate and return
+	if n := alloc(); n != nil {
+		return n, nil
+	}
+
+	// try expanding current pool
+	if err := pa.Expand(memCls); err != nil {
+		// this means there's not enough space left
+		if err == ErrInvalidMemoryClass {
+			return nil, ErrNotEnoughSpace
+		}
+		return nil, err
+	}
```

Comment:
Should we also try to merge if expand doesn't work and then error out if that fails? That may be overcomplicated for the use case for this though

Reply:
with the current implementation it's not that straightforward as you pointed out... I could give it a shot by changing the Expand "api" or moving the logic to Allocate, this way we can have a track of what needs to be merged back.